pyexchange/erisx.py

```diff
@@ -0,0 +1,400 @@
+# This file is part of Maker Keeper Framework.
+#
+# Copyright (C) 2019 EdNoepel
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import json
+import jwt
+import logging
+import requests
+import simplefix
+import time
+import uuid
+import datetime
+
+from dateutil import parser
+from typing import List
+
+from pyexchange.api import PyexAPI
+from pyexchange.fix import FixEngine
+from pyexchange.model import Order, Trade
+from pymaker.util import http_response_summary
+from pymaker.numeric import Wad
+
+
+class ErisxOrder(Order):
+
+    @staticmethod
+    def from_message(item: dict) -> Order:
+        return Order(order_id=item['oid'],
+                     timestamp=item['created_at'],
+                     pair=item['book'],
+                     is_sell=True if item['side'] == 'sell' else False,
+                     price=Wad.from_number(item['price']),
+                     amount=Wad.from_number(item['amount']))
+
+
+class ErisxTrade(Trade):
+
+    @staticmethod
+    def from_message(trade: dict) -> Trade:
+        return Trade(trade_id=trade['trade_id'],
+                     timestamp=int(parser.isoparse(trade['time']).timestamp()),
+                     pair=trade["contract_symbol"],
+                     is_sell=True if trade['side'] == 'SELL' else False,
+                     price=Wad.from_number(trade['px']),
+                     amount=Wad.from_number(abs(float(trade['qty']))))
+
+
+class ErisxApi(PyexAPI):
+    """
+    Implementation logic for interacting with the ErisX exchange, which uses FIX for order management and
+    market data, and a WebAPI for retrieving account balances.
+
+    ErisX documentation available here: https://www.erisx.com/wp-content/uploads/2020/03/ErisX-FIX-4.4-Spec-V3.3.pdf
+    """
+
+
+    logger = logging.getLogger()
+    timeout = 5
+
+    def __init__(self, fix_trading_endpoint: str, fix_trading_user: str,
+                 fix_marketdata_endpoint: str, fix_marketdata_user: str, password: str,
+                 clearing_url: str, api_key: str, api_secret: str, web_api_only: bool):
```

Comment:
Should `web_api_only` be implemented by letting the user pass `clearing_url=None`?  Note this seems to have broken `manual_test_erisx.py`.

Reply:
ah, i only updated `manual_test_erisx.py` locally. Thanks for the suggestion, that would be much cleaner. In the same vein, could perhaps use `""` instead of `None` so we could preserve the `isinstance` checks? And then check against that and only pass non empty string if needed

Reply:
Although they're falsey, I'm not a fan of empty strings.  We usually do `assert isinstance(endpoint, str) or endpoint is None`.